### PR TITLE
bug #478: decode html entities in form title

### LINKED ISSUE #478: Form titles with `&` not decoded properly
If you put an `&` in the XLSForm title, Central’s UI/DB shows `&amp;` while Enketo and Collect show the correct `&`.

*xlsform*
xlsform

*central ui (and db)*
central

*xml*
Screen Shot 2022-05-05 at 09 42 21

*enketo*
enketo

@matthew-white says...
> Frontend just uses what it gets from the API, so I think it’d be an issue on central-backend. We probably just need to decode the text after we extract it from the XML.

If we don't already have tests, it'd also be good to add some tests for Unicode characters. 

Potentially relevant: I vaguely remember that HTML documents don't support Unicode in the `<title>` tag.

### PR DESCRIPTION
**Pull request by sadiqkhoja**
Closes #478 

### Change
I have decoded only title before saving it in database

### Verification
Created unit tests
Manually tested in ODK Collect and Enketo to ensure that change is not breaking functionality there

### Alternative
I could have decoded the title in GET APIs after retrieving the data from DB but there are few of them. 
Other option was to change the `text()` function but I am afraid of ripple effect it might create across the application

### CODE REVIEW

File: lib/util/util.js [hunk 1]
```diff
@@ -33,6 +33,13 @@ const sanitizeOdataIdentifier = (name) =>
   name.replace(/^([^a-z_])/i, '_$1').replace(/([^a-z0-9_]+)/gi, '_');
 
 
+// Decodes html entities into plain text
+// greedy approach, only converting entities that are return by xlsform to xform conversion process
+const decodeHtml = (input) =>
+  input.replace(/&amp;/g, '&')
+    .replace(/&lt;/g, '<')
+    .replace(/&gt;/g, '>');
```
Comment (matthew-white) [inline lib/util/util.js:41]: Elsewhere we use this `decodeXML()` function:

https://github.com/getodk/central-backend/blob/321ac47d182e64cfffa9fb9e4ed421c5a614cb4a/lib/data/schema.js#L24

Would it be possible to use that instead of writing our own `decodeHtml()` function?

File: package-lock.json [hunk 2]
```diff
@@ -769,7 +769,7 @@
     "append-field": {
       "version": "1.0.0",
       "resolved": "https://registry.npmjs.org/append-field/-/append-field-1.0.0.tgz",
-      "integrity": "sha512-klpgFSWLW1ZEs8svjfb7g4qWY0YS5imI82dTg+QahUvJ8YqAY0P10Uk8tTyh9ZGuYEZEMaeJYCF5BFuX552hsw=="
+      "integrity": "sha1-HjRA6RXwsSA9I3SOeO3XubW0PlY="
```
Comment (matthew-white) [inline package-lock.json:772]: Is it intentional that this changed?

**Comment by matthew-white** [comment]
I suspect that there are other places where we should decode XML text. For example, I'm guessing we should decode the submission instance name:

https://github.com/getodk/central-backend/blob/321ac47d182e64cfffa9fb9e4ed421c5a614cb4a/lib/model/frames/submission.js#L36-L62

XML text isn't automatically decoded. I'm not sure, but I believe that XML attributes also aren't. If that's the case, there are probably places where we should decode XML attributes (for example, the form version string).

That said, I think it's reasonable to fix the specific issue filed. @lognaturel, let me know if you want me to file a separate issue about investigating other places to decode XML text and attributes.

File: package-lock.json (hunk 2 above)
Comment (sadiqkhoja) [reply package-lock.json:772]: I ran `npm install` and it modified it.

Comment (matthew-white) [reply package-lock.json:772]: Interesting! I wonder what causes `integrity` to change like that? 🤔

Since it's a separate change from this issue, what do you think about creating a new PR for that particular change?

Comment (sadiqkhoja) [reply package-lock.json]: done

**Comment by lognaturel** [comment]
> for example, the form version string

Didn't we already make a fix specifically for this?

I don't think we need to file issues to look for other places to do this. Starting with Collect v2022.3, there will be no more HTML encoding of unicode so these kinds of issues would naturally phase out anyway.

**Comment by matthew-white** [comment]
> > for example, the form version string
> 
> Didn't we already make a fix specifically for this?

I don't quite remember. Did we make a change to support Unicode? But it does look like we currently don't decode `&amp;`: https://test.getodk.cloud/#/projects/388/forms/ampersand_version

> Starting with Collect v2022.3, there will be no more HTML encoding of unicode so these kinds of issues would naturally phase out anyway.

Even without encoding Unicode, I think we'll still have issues with `&`, `<`, and `>`. Maybe those aren't super common characters though?

**Comment by matthew-white** [comment]
Oh and since this came up when @sadiqkhoja and I chatted, I also wanted to mention explicitly that this fix won't undertake a backfill: this will be a fix going forward for newly published forms and form versions.

**Comment by lognaturel** [comment]
> Did we make a change to support Unicode? But it does look like we currently don't decode &amp;: https://test.getodk.cloud/#/projects/388/forms/ampersand_version

Right, sorry, I conflated two things! Agreed they aren't super common. I think my preference is to wait and see if anyone complains so we can learn more about why they're putting those characters in version or formid strings.

**Review by matthew-white — APPROVED** [review]

**Comment by matthew-white** [comment]
Oh wait, don't we now need to encode the form def name in the OpenRosa `formList`? For example, if the form title is `&amp;`, it should show up in the formList as `<name>&amp;amp;</name>`, not `<name>&amp;</name>`: I think the latter would be shown as `&`, not the original `&amp;`.

Or is this not really a problem in practice?

**Comment by lognaturel** [comment]
> if the form title is &amp;

Well, more importantly, if the title is Crate & Barrel, wouldn't the form list have invalid XML of `<name>Crate & Barrel</name>`? So yes, this seems like it should be accompanied by a corresponding encode unfortunately.

**Comment by sadiqkhoja** [comment]
When form title is `Crate & Barrel` in database, openRosa formList shows it as `<name>Crate &amp; Barrel</name>`.

**Comment by lognaturel** [comment]
> formList shows it as <name>Crate &amp; Barrel</name>.
Great, then. For Matt's example, if you put &amp; as the title in an XLSForm, it will be &amp;amp; in the XForm, &amp; in the db and &amp;amp; in the formList?

Doesn't that mean there's an encode somewhere? Was it double-encoding before? (e.g. "Crate &amp;amp; Barrel" in formList?)

**Comment by sadiqkhoja** [comment]
It does double encoding. 

Fill blank form show title correctly, probably it's reading it from xml form definition:
Screenshot_20220729-145830

Get blank form shows `&amp;` in the title, probably reading from `name` field in formList response
Screenshot_20220729-145816

**Comment by lognaturel** [comment]
Great, so we get to fix 2 issues with this one. Looks ready to go to me.

**Comment by matthew-white** [comment]
Just as a quick check (and to prevent any future regression), how about we add a test that the OpenRosa `formList` encodes the form title/name?

**Review by matthew-white — APPROVED** [review]
Thanks, @sadiqkhoja! It's great to have that test in there. I'm thinking that Mustache must be escaping for us.
